std/algorithm/comparison.d

```diff
@@ -1567,3 +1567,156 @@ unittest
                 2, "two",
                 ));
 }
+
+/** Returns: First Argument (element of $(D a)) whose implicit conversion to
+    $(D bool) is $(D true.
+
+    Similar to behaviour of or operator in dynamic languages such as Lisp's (or
+    a...) and Python's a or ....
+
+    TODO: Is inout Conversion!T the correct return value?
+
+    NOTE: Lazy parameters are currently, too restrictively, inferred by DMD to
+    always throw eventhough they don't need to be. This makes it impossible to
+    mark $(D either) as $(D nothrow). Issue at
+    https://issues.dlang.org/show_bug.cgi?id=12647
+*/
+CommonType!Ts either(Ts...)(lazy Ts a)
+if (a.length >= 1 && !is(CommonType!Ts == void))
+{
+    foreach (e; a)
+        if (e)
+            return e;
+    return CommonType!Ts.init;
+}
+
+/// ditto
+auto ref either(Ts...)(ref Ts a)
+if (a.length >= 1 && allSameType!Ts)
+{
+    foreach (ref e; a[0 .. $ - 1])
+        if (e)
+            return e;
+    return a[$ - 1];
+}
+
+///
+@safe pure unittest
+{
+    immutable p = 1, q = 2;
+
+    auto pq = either(p, q);
+    assert(pq == 1);
+
+    assert(either(3) == 3);
+    assert(either(3, 4) == 3);
+    assert(either(0, 4) == 4);
+    assert(either(0, 0) == 0);
+    assert(either("", "a") == "");
```

Comment:
You should make sure to point out this case as it's not intuitive for someone who doesn't know how D's arrays work.